### DISCUSSION TITLE
Why List<Double> instead of float[] for embeddings ?

### DISCUSSION THREAD

**Discussion by michaelsembwever** [spring-projects/spring-ai#632] [General]
Curious to why `List<Double>` was used, considering the performance benefits a primitive array gives, particularly with SIMD operations newer Java versions can do.

A number of the vector stores have already a `toFloatArray(List<Double>)` to workaround this.

Bonus question :) 
What about using the [JVector](https://github.com/jbellis/jvector) library and its `VectorFloat` class …? 
It's heavily optimised against the Panama Vector API(SIMD), and detects and takes advantage of JDK20+ when available.

While spring-ai might not be doing much with the embeddings, it's quite likely consumers of the framework will be, e.g. re-ranking large similarity result sets.

**Comment by ThomasVitale** [comment]
This was addressed in version M2, where the type of embeddings has been changed from `List<Double>` to `float[]`. See: https://github.com/spring-projects/spring-ai/pull/1002
